concurrent/src/main/java/com/facebook/airlift/concurrent/ConcurrentScheduledExecutor.java

```diff
@@ -0,0 +1,166 @@
+
+    public ConcurrentScheduledExecutor(int schedulerCount, int threadsPerScheduler, String threadNameFormat)
+    {
+        checkArgument(schedulerCount > 0, "schedulerCount must be at least one");
+        checkArgument(threadsPerScheduler > 0, "threadsPerScheduler must be at least one");
+        ThreadFactory threadFactory = daemonThreadsNamed(requireNonNull(threadNameFormat, "threadBaseName is null"));
+
+        ImmutableList.Builder<ScheduledExecutorService> schedulersBuilder = ImmutableList.builder();
+        for (int i = 0; i < schedulerCount; i++) {
+            ScheduledThreadPoolExecutor scheduledExecutorService = new ScheduledThreadPoolExecutor(threadsPerScheduler, threadFactory);
+            scheduledExecutorService.setRemoveOnCancelPolicy(true);
+            schedulersBuilder.add(scheduledExecutorService);
+        }
+        schedulers = schedulersBuilder.build();
+    }
+
+    public ScheduledExecutorService getScheduledExecutorService()
+    {
+        return schedulers.get(ThreadLocalRandom.current().nextInt(schedulers.size()));
+    }
+
+    @Override
+    public ScheduledFuture<?> schedule(Runnable command, long delay, TimeUnit unit)
+    {
+        return getScheduledExecutorService().schedule(command, delay, unit);
+    }
+
+    @Override
+    public <V> ScheduledFuture<V> schedule(Callable<V> callable, long delay, TimeUnit unit)
+    {
+        return getScheduledExecutorService().schedule(callable, delay, unit);
+    }
+
+    @Override
+    public ScheduledFuture<?> scheduleAtFixedRate(Runnable command, long initialDelay, long period, TimeUnit unit)
+    {
+        return getScheduledExecutorService().scheduleAtFixedRate(command, initialDelay, period, unit);
+    }
+
+    @Override
+    public ScheduledFuture<?> scheduleWithFixedDelay(Runnable command, long initialDelay, long delay, TimeUnit unit)
+    {
+        return getScheduledExecutorService().scheduleWithFixedDelay(command, initialDelay, delay, unit);
+    }
+
+    @Override
+    public List<Runnable> shutdownNow()
+    {
+        return schedulers.stream()
+                .map(s -> s.shutdownNow())
+                .flatMap(list -> list.stream())
+                .collect(Collectors.toList());
+    }
+
+    @Override
+    public boolean isShutdown()
+    {
+        return schedulers.stream().allMatch(s -> s.isShutdown());
+    }
+
+    @Override
+    public boolean isTerminated()
+    {
+        return schedulers.stream().allMatch(s -> s.isTerminated());
+    }
+
+    @Override
+    public boolean awaitTermination(long timeout, TimeUnit unit)
+            throws InterruptedException
+    {
+        long start = System.currentTimeMillis();
+        long timeoutMillis = unit.toMillis(timeout);
+        boolean awaitResult = true;
+        for (ScheduledExecutorService executor : schedulers) {
+            awaitResult &= executor.awaitTermination(timeout, unit);
+            if (!awaitResult || System.currentTimeMillis() - start > timeoutMillis) {
```

Comment:
I don't think it is correct. Even if the timeout is gone, it doesn't mean that the remaining schedulers have not finished.